Python/ceval.c

```diff
@@ -2062,6 +2059,100 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
             DISPATCH();
         }
 
+        // TODO: Merge tails of *_ADD and BINARY_ADD.
+        // Alternatively, remove the INCREF/DECREF(right) pair,
+        // since the local keeps them alive and there's
+        // no way the local can be deleted by PyNumber_Add()
+        // (if there was a call to a closure with access, 
+        // it would be a cell, not a fast local).
+        // Merging is inconvenient because of scopes, though.
+        case TARGET(FAST_ADD): {
+            PyObject *right = GETLOCAL(oparg);
+            if (right == NULL) {
+                goto undefined_fast_local;
+            }
+            // Py_INCREF(right);
+            PyObject *left = TOP();
+            PyObject *sum;
+            if (PyUnicode_CheckExact(left) &&
+                     PyUnicode_CheckExact(right)) {
+                sum = unicode_concatenate(tstate, left, right, f, next_instr);
+                /* unicode_concatenate consumed the ref to left */
+            }
+            else {
+                sum = PyNumber_Add(left, right);
+                Py_DECREF(left);
+            }
+            // Py_DECREF(right);
+            SET_TOP(sum);
+            if (sum == NULL)
+                goto error;
+            DISPATCH();
+        }
+
+        // See comments at FAST_ADD
+        case TARGET(CONST_ADD): {
+            PyObject *right = GETITEM(consts, oparg);
+            // Py_INCREF(right);
+            PyObject *left = TOP();
+            PyObject *sum;
+            if (PyUnicode_CheckExact(left) &&
+                     PyUnicode_CheckExact(right)) {
+                sum = unicode_concatenate(tstate, left, right, f, next_instr);
+                /* unicode_concatenate consumed the ref to left */
+            }
+            else {
+                sum = PyNumber_Add(left, right);
+                Py_DECREF(left);
+            }
+            // Py_DECREF(right);
+            SET_TOP(sum);
+            if (sum == NULL)
+                goto error;
+            DISPATCH();
+        }
+
+        case TARGET(INT_ADD): {
+            PyObject *left = TOP();
+            if (PyLong_CheckExact(left)) {
+                PyLongObject *ll = (PyLongObject *)left;
```

Comment:
I'd move this code into a `PyObject *_PyLong_AddInt(PyLongObject *left, int right)` inline function in longobject.h
It will make life easier if we want to improve the implementation of `PyLongObject`.